spec/src/main/asciidoc/core/spi_lite.asciidoc

```diff
@@ -1,12 +1,386 @@
+
+[source, java]
+----
+public interface BuildCompatibleExtension {}
+----
+
+Build compatible extensions can define arbitrary `public`, non-`static`, `void`-returning methods without type parameters, annotated with one of the extension annotations.
+Such methods are called _extension methods_.
+
+Extension annotations correspond to extension execution phases:
+
+* `@Discovery`
+* `@Enhancement`
+* `@Registration`
+* `@Synthesis`
+* `@Validation`
+
+Extension methods can declare arbitrary number of parameters.
+Which parameters can be declared depends on the particular execution phase.
+All the parameters will be provided by the container when the extension method is invoked.
+
+The container instantiates a single instance of each build compatible extension and invokes all extension methods on this instance.
+
+The invocation order for extension methods may be declared using the `@Priority` annotation.
+If an extension method does not have the `@Priority` annotation, the default priority `jakarta.interceptor.Interceptor.Priority.APPLICATION + 500` is assumed.
+Note that priority only affects order of extension methods in a single phase.
+
+If an extension method throws an exception, the container treats it as a deployment problem.
+
+At deployment time, CDI container does not have to be running, so calling `CDI.current()` from an extension method, or attempting to access a running CDI container in any other way, results in non-portable behavior.
+
+[[bce_discovery]]
+
+=== The `@Discovery` phase
+
+In this phase, build compatible extensions may register additional classes to be scanned during type discovery, and register custom CDI meta-annotations.
+
+Extension methods annotated `@Discovery` may declare parameters of the following types:
+
+* `ScannedClasses`
+* `MetaAnnotations`
+* `Messages`
+
+[source, java]
+----
+public interface ScannedClasses {
+    void add(String className);
+}
+----
+
+[source, java]
+----
+public interface MetaAnnotations {
+    ClassConfig addQualifier(Class<? extends Annotation> annotation);
+    ClassConfig addInterceptorBinding(Class<? extends Annotation> annotation);
+    ClassConfig addStereotype(Class<? extends Annotation> annotation);
+
+    void addContext(Class<? extends Annotation> scopeAnnotation, Class<? extends AlterableContext> contextClass);
+    void addContext(Class<? extends Annotation> scopeAnnotation, boolean isNormal, Class<? extends AlterableContext> contextClass);
+}
+----
+
+If the `addQualifier`, `addInterceptorBinding` or `addStereotype` method is called, the return value allows configuring meta-annotations on the qualifier, interceptor binding or stereotype annotation and its members.
+
+[source, java]
+----
+public interface Messages {
+    void info(String message);
+    void info(String message, AnnotationTarget relatedTo);
+    void info(String message, BeanInfo relatedTo);
+    void info(String message, ObserverInfo relatedTo);
```

Comment:
Hmm, you cannot get hold of `BeanInfo` at this point (and `ObserverInfo`). I understand you are using the same class for multiple phases but this is rather confusing. I'd probably just say that `Messages` class should be used for logging or throwing an error and leave it at that without showing its code.
Or perhaps show it separately at a later stage once you introduce all  phases?